### PR TITLE
Relayer

### PR DESCRIPTION
**Pull request by ana0**

### CODE REVIEW

File: src/server/blockchain/index.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const Web3 = require('web3');
```
Comment (adzialocha) [inline src/server/blockchain/index.js]: Is it possible to use `import` syntax here (for constistency) or does some other part require this syntax?

File: src/server/blockchain/config.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require('dotenv').config();
```
Comment (adzialocha) [inline src/server/blockchain/config.js]: The environment variables were already set in the main file here: https://github.com/lazaruslabs/culturestake/blob/master/src/server/index.js#L13 - there is not need to call the function twice.

File: src/server/blockchain/config.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+
+const rpcUrl = process.env.RPC_URL || 'http://localhost:8545';
+
+module.exports = {
```
Comment (adzialocha) [inline src/server/blockchain/config.js]: Similar here: Could we stick to `export` syntax for consistency?

File: src/server/blockchain/index.js (hunk 1 above)
Comment (adzialocha) [reply src/server/blockchain/index.js]: We have `web3` already defined here: https://github.com/lazaruslabs/culturestake/blob/master/src/common/services/web3.js is it possible to extend the needed functionality there?

File: src/server/blockchain/config.js (hunk 2 above)
Comment (ana0) [reply src/server/blockchain/config.js]: I think I was echoing what's in `database/config.js`, but can remove

File: src/server/blockchain/config.js (hunk 3 above)
Comment (ana0) [reply src/server/blockchain/config.js]: Same as ^

File: src/server/blockchain/index.js (hunk 1 above)
Comment (ana0) [reply src/server/blockchain/index.js]: Yup, didn't realize this was here

File: src/server/blockchain/config.js (hunk 2 above)
Comment (adzialocha) [reply src/server/blockchain/config.js]: @ana0 I see! In database we have to do it like that as the file is imported by the `sequelize-cli` command line application which does not support `import` syntax.

File: src/common/services/contracts.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * Helper method to get a deployed smart contract instance.
+ *
+ * @param {Web3} web3 - Web3 instance
+ * @param {Object} abi - contract abi
+ * @param {Object} address - contract address
+ *
+ * @return {Object} - contract instance
+ */
+function getContract(web3, abi, address) {
```
Comment (adzialocha) [inline src/common/services/contracts.js]: We can get the `web` instance through an import here. It was not possible in the Circles implementation you copied the code from :smile_cat:

File: src/common/services/contracts.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import RelayerContract from 'culturestake-contracts/build/contracts/Relayer.json';
+import AdminContract from 'culturestake-contracts/build/contracts/Admin.json';
+import VoteContract from 'culturestake-contracts/build/contracts/Vote.json';
+
+/**
+ * Helper method to get a deployed smart contract instance.
+ *
+ * @param {Web3} web3 - Web3 instance
```
Comment (adzialocha) [inline src/common/services/contracts.js]: Looks funny with these JSDoc comments when there are none anywhere else, I'm fine with removing them :+1:

File: src/common/services/web3.js [hunk 6]
```diff
@@ -1,5 +1,12 @@
...
 
-export default web3;
+const provider = new Web3.providers.WebsocketProvider(rpcUrl);
+const web3 = new Web3(provider);
+
+const connect = async () => {
+  return (await web3.eth.getBlock('latest')).number;
+};
+
+export default { connect, web3, provider };
```
Comment (adzialocha) [inline src/common/services/web3.js]: You can add the export statements directly where the objects are defined, like:

`export default const web3 = ...`
`export function connect = ...`

The provider one can always retreive via https://web3js.readthedocs.io/en/v1.2.6/web3-bzz.html#currentprovider

Also I would maybe rename `connect` as the method is not "really" connecting to the blockchain (I understand `connect` as an required step to initialize a whole system while here it is used to just check if there IS a connection).

File: src/server/controllers/relay.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import { respondWithSuccess } from '~/server/helpers/respond';
+import metaTxHandler from '~/server/services/metatx';
+import blockchain from '~/common/services/web3';
```
Comment (adzialocha) [inline src/server/controllers/relay.js]: With the above comment you can write now `import web3 from ...`

File: src/server/routes/relay.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import express from 'express';
+
+import relayController from '~/server/controllers/relay';
+import relayValidation from '~/server/validations/relay';
+import validate from '~/server/helpers/validate';
+
+const router = express.Router();
+
+//router.post('/', () => {console.log('middle')}, validate(relayValidation.metatx), () => {console.log('middle2')}, relayController.relay);
```
Comment (adzialocha) [inline src/server/routes/relay.js]: No "dead comments" please :smiley_cat:

File: test/relay.test.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      relayer.options.jsonInterface,
+    );
+  });
+
+  describe('POST /api/relay', () => {
+    it('should respond with a successful message', async () => {
+      const metaNonce = await relayer.methods
+        .getNonce(sender.address.toString())
+        .call();
+      // console.log(metaNonce)
```
Comment (adzialocha) [inline test/relay.test.js]: Same here!

File: test/relay.test.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        nonce: metaNonce,
+        data: vote.methods.recordVote(sender.address).encodeABI(),
+      };
+      const metaSignedTx = await metaTxHandler.signMetaTx(
+        txParams,
+        sender.privateKey.substring(2),
+        metaNonce,
+      );
+      // console.log(sender.address)
+      // console.log('signed', metaSignedTx)
```
Comment (adzialocha) [inline test/relay.test.js]: ... and here!

File: src/common/services/contracts.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import RelayerContract from 'culturestake-contracts/build/contracts/Relayer.json';
+import AdminContract from 'culturestake-contracts/build/contracts/Admin.json';
+import VoteContract from 'culturestake-contracts/build/contracts/Vote.json';
+
+import web3 from './web3';
```
Comment (adzialocha) [inline src/common/services/contracts.js]: This can be changed to absolute paths as `import web3 from '~/common/services/web3';`

**Pull request closed without merging**